draft-ietf-webpush-encryption.md

```diff
@@ -108,170 +101,277 @@ encryption scheme by applications that use push messaging.  An agent that only
 
-The Application Server generates a new DH or ECDH key pair in the same group as
-the value generated by the User Agent.
+L:
+: 32 octets (i.e., the output is the length of the underlying SHA-256 HMAC
+  function output)
 
-From the newly generated key pair, the Application Server performs a DH or ECDH
-computation with the public key provided by the User Agent to find the input
-keying material for key derivation.  The Application Server then generates 16
-octets of salt that is unique to the message.  A random [RFC4086] salt is
-acceptable.
 
-Web push uses the authentication secret defined in Section 4.3 of
-[I-D.ietf-httpbis-encryption-encoding].  This authentication secret (see
-{{auth}}) is generated by the user agent and shared with the application server.
+## Key Derivation Context {#context}
 
+The derivation of the content encryption key and nonce uses an additional
+context string.
+
+The context is comprised of a label of "P-256" encoded in ASCII (that is, the
+octet sequence 0x50, 0x2d, 0x32, 0x35, 0x36), a zero-valued octet, the length of
+the User Agent public key (65 octets) encoded as a two octet unsigned integer in
+network byte order, the User Agent public key, the length of the Application
+Server public key (65 octets), and the Application Server public key.
+
+~~~ inline
+   context = label || 0x00 ||
+               length(ua_public) || ua_public ||
+               length(as_public) || as_public
+~~~
 
-## Push Message Content Encryption {#c-e}
+## Encryption Summary {#summary}
+
+This results in a the final content encryption key and nonce generation using
+the following sequence, which is shown here in pseudocode with HKDF expanded
+into separate discrete steps using HMAC with SHA-256:
+
+~~~ inline
+   -- For a User Agent:
+   ecdh_secret = ECDH(ua_private, as_public)
+   auth_secret = random(16)
+
+   -- For an Application Server:
+   ecdh_secret = ECDH(as_private, ua_public)
+   auth_secret = <from User Agent>
+
+   -- For both:
+   auth_info = "Content-Encoding: auth" || 0x00
+   PRK_combine = HMAC-SHA-256(auth_secret, ecdh_secret)
+   IKM = HMAC-SHA-256(PRK_combine, auth_info || 0x01)
+   context = "P-256" || 0x00 ||
+             0x00 || 0x41 || ua_public ||
+             0x00 || 0x41 || as_public
+   salt = random(16)
+   PRK = HMAC-SHA-256(salt, IKM)
+   cek_info = "Content-Encoding: aesgcm" || 0x00 || context
+   CEK = HMAC-SHA-256(PRK, cek_info || 0x01)[0..15]
+   nonce_info = "Content-Encoding: nonce" || 0x00 || context
+   NONCE = HMAC-SHA-256(PRK, nonce_info || 0x01)[0..11]
+~~~
 
-The Application Server then encrypts the payload.  Header fields are populated
-with base64url encoded [RFC7515] values:
+Note that this omits the exclusive OR of the final nonce with the record
+sequence number, since implementations of this specification are not required to
+support multiple records (see {{restrict}}) and the sequence number of the first
+record is zero.
```

Comment:
> since implementations of this specification are not required to support multiple records

Is contradictory to the language in restrictions:

> An Application Server MUST encrypt a push message with a single record.

I suggest replacing the former with "since there must be at most a single record (see {{restrict}})."